### PR TITLE
Refactor and migration to nf-core

### PR DESCRIPTION
**Pull request by mberacochea**

### CODE REVIEW

File: bin/binlinks.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 import csv
 import gzip
 
+# FIXME: why is this script here and we are not using the one in eukcc?
```
Comment (KateSakharova) [inline bin/binlinks.py:12]: I didn't know Paul added it to eukcc container

File: config/nf_modules.config [hunk 2]
```diff
@@ -102,20 +114,20 @@ process {
...
         publishDir = [
             [
-                path: { "${params.outdir}/intermediate_steps/Binning/CONCOCT/bins" },
+                path: { "${params.outdir}/intermediate_steps/binning/concot/bins" },
                 mode: params.publish_dir_mode,
                 saveAs: { filename -> new File(filename).getName() },
                 pattern: "*/*.fa"
             ]
         ]
-        ext.prefix = { "CONCOCT-${meta.id}" }
```
Comment (KateSakharova) [inline config/nf_modules.config:118]: That should be CONCOCT as it is name of [tool](https://github.com/BinPro/CONCOCT) and we will grep name of the tool from bin filename and submit those values to ENA.

File: config/nf_modules.config [hunk 3]
```diff
@@ -192,59 +204,71 @@ process {
                         if ( !filename.endsWith(".out") ) {
                             return null;
                         }
-                        def output_file = file(filename);
+                        def output_file = new File(filename);
                         def genome_id = fasta.baseName;
                         if ( output_file.name.contains("_rRNAs") || output_file.name.contains("_tRNA_20aa") ) {
-                            return "${genome_id}/${genome_id}_out-results/${output_file.name}";
+                            return "${genome_id}/RNA/${genome_id}_out/${output_file.name}";
```
Comment (KateSakharova) [inline config/nf_modules.config:210]: I'm sure MAG_Uploader will not work with extra RNA folder level. Old script https://github.com/EBI-Metagenomics/assembly-pipeline/blob/develop/pipeline_lib/path_finder.py#L108

File: config/nf_codon_dbs.config [hunk 4]
```diff
@@ -5,15 +5,24 @@
...
+
+    erz_to_err_mapping_file = "/hps/nobackup/rdf/metagenomics/service-team/users/kates/eukrecover2/Vaginal/rename.tsv"
+
+    checkm2_db = "/hps/nobackup/rdf/metagenomics/service-team/ref-dbs/checkm2/CheckM2_database/CheckM2_database/uniref100.KO.1.dmnd"
+
+    gunc_db = "/hps/nobackup/rdf/metagenomics/service-team/production/ref-dbs/genomes-pipeline/gunc_db_2.0.4.dmnd"
+    gtdbtk_db = "/hps/nobackup/rdf/metagenomics/service-team/production/ref-dbs/genomes-pipeline/release207_v2"
+
+    busco_db = "/hps/nobackup/rdf/metagenomics/service-team/ref-dbs/busco_new"
+    eukcc_db = "/hps/nobackup/rdf/metagenomics/service-team/ref-dbs/eukccdb/eukcc2_db_ver_1.1"
```
Comment (KateSakharova) [inline config/nf_codon_dbs.config:20]: we should switch to version 1.2 (located in the same folder)

File: modules/local/gtdbtk/main.nf [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
+        "-v ${gtdbtk_refdata}:/opt/gtdbtk_refdata" }"
 
     input:
     path(genomes_fna, stageAs: "genomes_dir/*")
-    path gtdbtk_refdata
+    path(gtdbtk_refdata)
 
     output:
-    path("Taxonomy")
+    path 'gtdbtk_results.tar.gz', emit: gtdbtk_output_tarball
```
Comment (KateSakharova) [inline modules/local/gtdbtk/main.nf:14]: That should require a fix in MAG_Uploader to support archive (I don't see GTDB-Tk result in test results folder)

File: modules/local/utils.nf [hunk 6]
```diff
@@ -88,7 +90,11 @@ process CHANGE_READS {
     export to_accession=\$(cat help_file | cut -f2)
     echo "\${from_accession} --> \${to_accession}"
 
-    change_reads.py --reads ${reads} -f \${from_accession} -t \${to_accession} --change_dots_to_underscores
+    gunzip ${reads}
```
Comment (KateSakharova) [inline modules/local/utils.nf:93]: maybe return sed command?

File: subworkflows/local/prok_mags_generation.nf [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    metabat_depth = collected_binners_and_depth.map { it -> it[4] }
+
+    // -- bin refinement //
+    BIN_REFINEMENT( collected_binners, checkm2_db )
+
+    // -- clean bins
+    CLEAN_AND_FILTER_BINS( 
+        BIN_REFINEMENT.out.bin_ref_bins,
+        cat_db_folder,
```
Comment (KateSakharova) [inline subworkflows/local/prok_mags_generation.nf:59]: what a magic, why no .first() 😮

**Review by KateSakharova — CHANGES_REQUESTED** [review]
Great work! We need to finish output directory review and merge PR.

File: config/nf_modules.config [hunk 8]
```diff
@@ -28,7 +38,7 @@ process {
                 path: "${params.outdir}/Taxonomy/Euks/",
```
Comment (KateSakharova) [inline config/nf_modules.config:38]: that should be renamed to `eukaryotes`

File: modules/local/coverage_recycler/main.nf
Comment (KateSakharova) [inline modules/local/coverage_recycler/main.nf:1]: That file creates an extra level of folders: publishDir does `coverage/prokaryotes` and script returns coverage folder. I think it we do expect `coverage/prokaryotes/<files>` only

File: subworkflows/local/euk_mags_generation.nf [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+    bins_alignment_by_bins = bins_alignment.map( spreadBins ).transpose(by: [1])  // tuple(meta, MAG1, [reads]); tuple(meta, MAG2, [reads])
+    
+    // TODO: leave only DREP_MAGS instead of DREP
+    ALIGN_BINS( bins_alignment_by_bins ) // out: [meta, fasta, bam, bai]
+
+    // ch_versions.mix( ALIGN_BINS.out.versions.first() )
+
+    // input: tuple(meta, MAG, bam, bai)
+    BREADTH_DEPTH( ALIGN_BINS.out.assembly_bam )
```
Comment (KateSakharova) [inline subworkflows/local/euk_mags_generation.nf:227]: We need to add a step to generate coverage per each MAG. MAG uploader needs file with 1 value per each MAG
example:
`SRR12830910_binner1.CONCOCT_v1.1.0-SRR12830910_bin.16.fa_coverage/SRR12830910_binner1.CONCOCT_v1.1.0-SRR12830910_bin.16.fa_MAGcoverage.txt`

```
SRR12830910_binner1.CONCOCT_v1.1.0-SRR12830910_bin.16.fa        0.7456465497737125
```

File: bin/binlinks.py (hunk 1 above)
Comment (mberacochea) [reply bin/binlinks.py:12]: it should be there.. it's in the repo at least -> https://github.com/EBI-Metagenomics/EukCC/blob/master/scripts/binlinks.py
and eukcc is on bioconda so there is biocontainer

File: config/nf_modules.config (hunk 3 above)
Comment (mberacochea) [reply config/nf_modules.config:219]: OK, we may need to adjust the MAG Uploader then. At this point I think we need to define a very clear structure for the outputs.

Also, I feel we should go with all lowercase folder and file names

File: config/nf_codon_dbs.config (hunk 4 above)
Comment (mberacochea) [reply config/nf_codon_dbs.config:20]: done

File: modules/local/gtdbtk/main.nf (hunk 5 above)
Comment (mberacochea) [reply modules/local/gtdbtk/main.nf:14]: OK. I'll collect the bits that will require us to change on the MAG uploader in a ticket

File: subworkflows/local/prok_mags_generation.nf (hunk 7 above)
Comment (mberacochea) [reply subworkflows/local/prok_mags_generation.nf:59]: yeah, for parameters that are folders or files we shouldn't be using `channel.from`, we should use `file`. We then can remove the `.first()`

File: config/nf_modules.config (hunk 8 above)
Comment (mberacochea) [reply config/nf_modules.config:38]: done

File: config/nf_modules.config [hunk 10]
```diff
@@ -10,13 +10,23 @@
...
+    }
+
     withName: ALIGNMENT {
         ext.alignment_args = [
             "-q", "20",
             "-Sb",
         ].join(' ').trim()
         ext.decontamination_args = [
-            "-f", "12",
+            "-f", "4",
```
[the remark this answers is not in the document]
Comment (Vkale1) [reply config/nf_modules.config:29]: Hmm.. so -f = keep reads that are... 4 = unmapped to host, 12 = read unmapped or mate unmapped to host (https://www.samformat.info/sam-format-flag) . 12 should work in theory... I would also add -F 256 (don't keep reads that are not a primary alignment).
Side note: did we veto bmtagger?

[the remark this answers is not in the document]
Comment (mberacochea) [reply config/nf_modules.config:29]: perfect, our code needs to account for pairs or not (which was causing the problem as we merge the reads and we were trying to use -f 12 so I changed it to -f 4). I'll refactor this as it needs to be integrated in the module itself and not the configuration. 
Thank you @Vkale1

**Comment by mberacochea** [comment]
I'm merging this one, and will record the items we discussed in JIRA. We should create features/bugfix branches to fix them.

File: modules/local/utils.nf (hunk 6 above)
Comment (KateSakharova) [reply modules/local/utils.nf:93]: we used `sed -i "s/_/\./" $F` in our commands before to change dots to underscores and can do the same with accessions

**Pull request merged**